File: Common Algorithm/Tree Traversal/Depth First Search/PostOrder Traversal/postorder.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def postorder_traversal(root):
    if not root:
        return []

    stack = [(root, False)]
    res = []

    while stack:
        node, visited = stack.pop()
        if visited:
            res.append(node.val)
        else:
            stack.append((node, True))
            if node.right:
                stack.append((node.right, False))
            if node.left:
                stack.append((node.left, False))

    return res

# Create a binary tree
#        1
#      /   \
#     2     3
#    / \   / \
#   4   5 6   7
root = TreeNode(1,
                TreeNode(2,
                         TreeNode(4),
                         TreeNode(5)),
                TreeNode(3,
                         TreeNode(6),
                         TreeNode(7)))

# Traverse the tree in post-order and print the values
result = postorder_traversal(root)
print(result)  # [4, 5, 2, 6, 7, 3, 1]

File: Common Algorithm/Tree Traversal/Depth First Search/PostOrder Traversal/postorder.js
class TreeNode {
    constructor(val, left, right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function postorderTraverse(root) {
    if (!root) return [];

    const stack = [[root, false]];
    const result = [];

    while (stack.length) {
        let [node, visited] = stack.pop();
        
        if (visited) result.push(node.val);
        else {

            stack.push([node, true]);
            if (node.right) stack.push([node.right, false]);
            if (node.left) stack.push([node.left, false]);

        }
    }

    return result;
}


const treeNode = new TreeNode(1,
        new TreeNode(2, new TreeNode(4), new TreeNode(5)),
        new TreeNode(3, new TreeNode(6), null)
  );
  
  const result = postorderTraverse(treeNode);
  console.log(result);